bindings/ocaml/EverCrypt.ml

```diff
@@ -0,0 +1,282 @@
+    | 0 -> Success (alg, st)
+    | n -> error n
+  let encrypt (alg, st) iv ad pt ct tag : unit result =
+    (* EverCrypt.AEAD.encrypt_pre *)
+    check_sizes alg iv tag;
+    assert (C.disjoint ct tag);
+    assert (C.disjoint iv ct);
+    assert (C.disjoint iv tag);
+    assert (C.disjoint pt tag);
+    assert (C.disjoint pt ad);
+    assert (C.disjoint ad ct);
+    assert (C.disjoint ad tag);
+    get_result (everCrypt_AEAD_encrypt (!@st)
+                  (C.ctypes_buf iv) (C.size_uint32 iv) (C.ctypes_buf ad) (C.size_uint32 ad)
+                  (C.ctypes_buf pt) (C.size_uint32 pt) (C.ctypes_buf ct) (C.ctypes_buf tag))
+  let decrypt (alg, st) iv ad ct tag dt : unit result =
+    (* EverCrypt.AEAD.decrypt_st *)
+    check_sizes alg iv tag;
+    assert (C.disjoint tag dt);
+    assert (C.disjoint tag ct);
+    assert (C.disjoint tag ad);
+    assert (C.disjoint ct ad);
+    assert (C.disjoint dt ad);
+    get_result (everCrypt_AEAD_decrypt (!@st)
+                  (C.ctypes_buf iv) (C.size_uint32 iv) (C.ctypes_buf ad) (C.size_uint32 ad)
+                  (C.ctypes_buf ct) (C.size_uint32 ct) (C.ctypes_buf tag) (C.ctypes_buf dt))
+end
+
+module Chacha20_Poly1305 : Chacha20_Poly1305 =
+  Make_Chacha20_Poly1305 (struct
+    let encrypt = EverCrypt_Chacha20Poly1305.everCrypt_Chacha20Poly1305_aead_encrypt
+    let decrypt = EverCrypt_Chacha20Poly1305.everCrypt_Chacha20Poly1305_aead_decrypt
+  end)
+
+module Curve25519 : Curve25519 =
+  Make_Curve25519 (struct
+    let secret_to_public = EverCrypt_Curve25519.everCrypt_Curve25519_secret_to_public
+    let scalarmult = EverCrypt_Curve25519.everCrypt_Curve25519_scalarmult
+    let ecdh = EverCrypt_Curve25519.everCrypt_Curve25519_ecdh
+  end)
+
+module Ed25519 : EdDSA =
+  Make_EdDSA (struct
+  let secret_to_public = EverCrypt_Ed25519.everCrypt_Ed25519_secret_to_public
+  let sign = EverCrypt_Ed25519.everCrypt_Ed25519_sign
+  let verify = EverCrypt_Ed25519.everCrypt_Ed25519_verify
+  let expand_keys = EverCrypt_Ed25519.everCrypt_Ed25519_expand_keys
+  let sign_expanded = EverCrypt_Ed25519.everCrypt_Ed25519_sign_expanded
+  end)
+
+module Hash = struct
+  open HashDefs
+  open EverCrypt_Hash
+
+  type t = alg * Z.t ref * everCrypt_Hash_Incremental_state_s ptr
+  let init alg =
+    let alg_spec = alg_definition alg in
+    let st = everCrypt_Hash_Incremental_create_in alg_spec in
+    everCrypt_Hash_Incremental_init st;
+    let incr_len = ref Z.zero in
+    (alg, incr_len, st)
+  let update (alg, incr_len, st) data =
+    assert (C.size data < max_input_len);
+    incr_len := Z.add !incr_len (Z.of_int (C.size data));
+    assert (Z.lt !incr_len (incremental_input_len alg));
+    everCrypt_Hash_Incremental_update st (C.ctypes_buf data) (C.size_uint32 data)
+  let finish (alg, _, st) dst =
+    assert (C.size dst = digest_len alg);
+    everCrypt_Hash_Incremental_finish st (C.ctypes_buf dst)
+  let free (_, _, st) =
+    everCrypt_Hash_Incremental_free st
```

Comment:
can this be attached as a finalizer on the OCaml GC block? would prevent misuses of the API